.github/workflows/unit-and-integration-tests-unified.yml

```diff
@@ -0,0 +1,160 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: Unit & Integration tests CI
+on:
+  push:
+    branches:
+      - master
+      - /^\d+\.\d+\.\d+(-\S*)?$/ # release branches
+  pull_request:
+    branches:
+      - master
+      - /^\d+\.\d+\.\d+(-\S*)?$/ # release branches
+
+concurrency:
+  group: '${{ github.workflow }} @ ${{ github.event.pull_request.head.label || github.head_ref || github.ref }}'
+  cancel-in-progress: true
+
+jobs:
+  build:
+    strategy:
+      fail-fast: false
+      matrix:
+        jdk: [ '8', '11', '17' ]
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout branch
+        uses: actions/checkout@v3
+      - name: Cache Maven m2 repository
+        id: maven
+        uses: actions/cache@v3
+        with:
+          path: ~/.m2/repository
+          key: maven-${{ runner.os }}-${{ matrix.jdk }}-${{ hashFiles('**/pom.xml') }}
+          restore-keys: |
+            maven-${{ runner.os }}-${{ matrix.jdk }}
+      - name: Cache targets
+        id: target
+        uses: actions/cache@v3
+        with:
+          path: |
+            ./target
+            ./extendedset/target
+            ./hll/target
+            ./processing/target
+            ./server/target
+            ./indexing-hadoop/target
+            ./web-console/target
+            ./indexing-service/target
+            ./services/target
+            ./benchmarks/target
+            ./distribution/target
+            ./integration-tests/target
```

Comment:
I wonder why target folders need to be cached, from what I gathered, the three types of tests running is doing following things roughly:
- unit tests: use mvn test (doesn't need packaging)
- it tests(old): use mvn verify (but invokes integration-tests profile and rebuilds entire modules anyhow - should this profile be modified to not build mvn all over again)
- it tests(new): reuses fixed docker image in build phase itself and only runs tests against this docker container.
Please correct me if I'm missing something here.

Reply:
Correction: we do need generated ` ./integration-tests-ex/image/target/env.sh` file and I updated caching to include all target folders, it would be nice to know why rest of the target folders need to be cached if I'm missing something.

Reply:
You highlight a number of questions. Yes, the `env.sh` file has to be cached between the image creation and the tests. It _would_ be possible to eliminate this file for GHA, if doing so was useful. The file is most helpful on the desktop when we want an easy developer experience. Let me know if you'd like the no-`env.sh` option: I can change the code to skip it (assuming all the required env vars are set.)

I'd need to check, but I suspect the only thing we use from `image/target` is the `env.sh` file. If we omit the file, as explained above, then there should be nothing else in that folder (as far as I remember) that we need to copy.

We can refine this in another PR so we can get this one in.